src/esphome/sensor/sds011_component.cpp

```diff
@@ -0,0 +1,258 @@
+// Based on:
+//   - https://cdn.sparkfun.com/assets/parts/1/2/2/7/5/Laser_Dust_Sensor_Control_Protocol_V1.3.pdf
+
+#include "esphome/defines.h"
+
+#ifdef USE_SDS011
+
+#include "esphome/log.h"
+#include "esphome/sensor/sds011_component.h"
+
+ESPHOME_NAMESPACE_BEGIN
+
+namespace sensor {
+
+static const char *TAG = "sensor.sds011";
+static const uint8_t SDS011_MSG_REQUEST_LENGTH = 19;
+static const uint8_t SDS011_MSG_RESPONSE_LENGTH = 10;
+static const uint8_t SDS011_DATA_REQUEST_LENGTH = 15;
+static const uint8_t SDS011_DATA_RESPONSE_LENGTH = 6;
+static const uint8_t SDS011_MSG_HEAD = 0xaa;
+static const uint8_t SDS011_MSG_TAIL = 0xab;
+static const uint8_t SDS011_COMMAND_ID_REQUEST = 0xb4;
+static const uint8_t SDS011_COMMAND_ID_RESPONSE = 0xc5;
+static const uint8_t SDS011_COMMAND_ID_DATA = 0xc0;
+static const uint8_t SDS011_COMMAND_REPORT_MODE = 0x02;
+static const uint8_t SDS011_COMMAND_QUERY_DATA = 0x04;
+static const uint8_t SDS011_COMMAND_SET_DEVICE_ID = 0x05;
+static const uint8_t SDS011_COMMAND_SLEEP = 0x06;
+static const uint8_t SDS011_COMMAND_FIRMWARE = 0x07;
+static const uint8_t SDS011_COMMAND_PERIOD = 0x08;
+static const uint8_t SDS011_GET_MODE = 0x00;
+static const uint8_t SDS011_SET_MODE = 0x01;
+static const uint8_t SDS011_MODE_REPORT_ACTIVE = 0x00;
+static const uint8_t SDS011_MODE_REPORT_QUERY = 0x01;
+static const uint8_t SDS011_MODE_SLEEP = 0x00;
+static const uint8_t SDS011_MODE_WORK = 0x01;
+
+SDS011Component::SDS011Component(UARTComponent *parent, uint32_t update_interval, bool query_mode, bool rx_mode_only)
+    : UARTDevice(parent),
+      update_interval_(update_interval),
+      query_mode_(query_mode),
+      rx_mode_only_(rx_mode_only),
+      force_query_(false) {
+}
+
+void SDS011Component::setup() {
+  if (this->get_rx_mode_only()) {
+    if (this->get_query_mode()) {
+      ESP_LOGE(TAG, "SDS011 component can not work in rx-only mode and query mode.");
+    }
+    return;
+  }
+
+  uint8_t command_data[SDS011_DATA_REQUEST_LENGTH] = {0};
+  command_data[0] = SDS011_COMMAND_REPORT_MODE;
+  command_data[1] = SDS011_SET_MODE;
+  command_data[2] = this->query_mode_ ? SDS011_MODE_REPORT_QUERY : SDS011_MODE_REPORT_ACTIVE;
+  command_data[13] = 0xff;
+  command_data[14] = 0xff;
+  this->sds011_write_command_(command_data);
+
+  uint8_t update_interval_minutes = this->get_update_interval() / 1000.0f / 60.0f;
+  this->set_update_interval((uint32_t)update_interval_minutes * 1000 * 60);
```

Comment:
The working period should be between 1-30 minutes and is given in minutes. (0 means continuous measurements). `update_interval_minutes` is sent to the sensor.  `update_interval` is only used to poll the sensor in query mode. . Casting in Python would be alright, but I do not know how to restrict the time type (in interval and maximum). Do you have an example?

Reply:
Ok, it's rather simple (I'll add it to the esphome pr directly). We should remove the code here, components shouldn't set their own properties - it probably works now but it could cause some trouble later.

Reply:
Thank you! I removed the casting.